### ISSUE TITLE
Dataset partitioning

### ISSUE THREAD

**Issue by wjones127** [lancedb/lance#952]
For design discussion, see [Design Proposal: File Partitions and Page Statistics in Lance Format](https://docs.google.com/document/d/1ovM2IhokkyJgM-rpLOZhw8yGo2LOmPkDIF7hOQ9D62E/edit)

## Basic partitioning support

* [ ] Add partition spec and partition values to the manifest file format
* [ ] Implement writing to partition table
* [ ] Add partition pruning to scans with filters
* [ ] Implement repartition table operation to rewrite for new spec

## Enhancements

* [ ] Add Bucket[N] partition transform
* [ ] Add date / timestamp partition transforms
* [ ] Add support for evolving partition schema
* [ ] Add per-partition KNN indices

**Comment by zero-element** [comment]
Any updates? Are there any plans and timelines for implementing partition_by?

**Comment by westonpace** [comment]
> Any updates? Are there any plans and timelines for implementing partition_by?

I don't believe it's currently a priority.  It's very useful for a few scenarios (mainly enabling range searches against the clustering key) but it is quite a bit of additional complexity (especially around compaction, updates, etc.) and it hasn't been crucial for us yet.
